### ISSUE TITLE
[BUG] IE11 -> Can't execute code from a freed script

### ISSUE THREAD

**Issue by lifaon74** [angular/zone.js#933]
**Context:**
- angular 5.0.0-rc3
- zone.js 0.8.18
- IE 11

I tried to update from zone 0.8.4 (current official angular 5 tutorials version) to 0.8.18, and get at some point with IE 11 the following error: `Can't execute code from a freed script`. This is a really imprecise error, that I didn't manage to find the origin but is due for sure from zone.js.

**Comment by JiaLiPassion** [comment]
@lifaon74 , could you provide a reproduce sample ?
or you can try to add following code in `index.html` before loading `zone.js`
```
<script>
__Zone_enable_cross_context_check = true;
</script>
```
to see the result.

**Comment by JiaLiPassion** [comment]
@lifaon74 , glad to help, and `__Zone_enable_cross_context_check ` this flag will enable some additional error check to `IE`, in `IE`, sometimes `DeveloperTools` will also register `eventListeners` which will cause some weird exceptions.

**Comment by ganeshkbhat** [comment]
Happening to me in edge also. Using the latest cli 1.5.4

**Comment by falcoprescher** [comment]
@lifaon74 While analyzing the migration of AngularJS to Angular 5 we came across the same problem you mentioned here with zone.js 0.8.18 in IE11. Downgrading it to 0.8.12 does not throw the exception anymore. Thank you for mentioning this workaround.

**Comment by JiaLiPassion** [comment]
@ganeshkbhat, @falcoprescher , could you guys try to add 
```javascript
<script>
__Zone_enable_cross_context_check = true;
</script>
```
in index.html before loading `zone.js`?

**Comment by falcoprescher** [comment]
@JiaLiPassion Setting this flag works which you suggested on the 20th October 2017. 
Is this a fix or a workaround?

**Comment by JiaLiPassion** [comment]
@falcoprescher , this is a fix and it should be better documented.
in old version, this flag is true by default, but it will have performance impact, so in recent version it has been changed to `false`, only in IE and some use cases it need to be set to true.

I will add document and better setting method later.

**Comment by falcoprescher** [comment]
@JiaLiPassion Thank you for the information. I am also considering disabling this flag in production mode via our webpack configuration.

**webpack.config.js**
```
new webpack.DefinePlugin({
    ...
    // Needs to be set for zone.js so it does not throw an exception in IE11 while being in developer tools
    // https://github.com/angular/zone.js/issues/933
    '__Zone_enable_cross_context_check': this.isDevMode ? 'true' : 'false'
}));
```
**app.base.ts**
```
import './app.globals.ts';
import 'zone.js';
...
```
**app.globals.ts**
```
declare let zoneEnableCrossContextCheck: boolean;
(<any>window).__Zone_enable_cross_context_check = zoneEnableCrossContextCheck;
```

**Comment by JiaLiPassion** [comment]
@falcoprescher , thank you for the configuration idea!

**Comment by fuerst** [comment]
How can that be configured in an Angular CLI generated project?

**Comment by ciesielskico** [comment]
Also having this problem in IE11

**Comment by JiaLiPassion** [comment]
@fuerst, @ciesielskico, you can just add the following line in `polyfill.ts`  before loading `zone.js`.
```javascript
__Zone_enable_cross_context_check = true;
import 'zone.js/dist/zone';  // Included with Angular CLI.
```

**Comment by shcallaway** [comment]
This might be a silly question, but the code above produces a compiler error for me:

`ERROR in src/polyfills.ts(58,1): error TS2304: Cannot find name '__Zone_enable_cross_context_check'.`

**Comment by JiaLiPassion** [comment]
@shcallaway , oh, sorry, please modify it to 

```javascript
(window as any).__Zone_enable_cross_context_check = true;
import 'zone.js/dist/zone';  // Included with Angular CLI.
```

**Comment by kevinkuszyk** [comment]
I'm seeing this intermittently even after adding the [fix mentioned above](https://github.com/angular/zone.js/issues/933#issuecomment-359686741).

I've pushed a sample repro at [kevinkuszyk/angular-ie-freed-script-error](https://github.com/kevinkuszyk/angular-ie-freed-script-error) which shows both this and #1001.

**Comment by danail-vasilev** [comment]
Hi @kevinkuszyk, can you try to set the property not in the polyfills file but directly in the index.html file? In my project when I set it in the index.html it is not reproducible but when it is set in the polyfills it doesn't take effect.
Thanks !

**Comment by JiaLiPassion** [comment]
@danail-vasilev ,
there will be websocket to connect angular web app with angular dev server, and every time refresh IE page, there will be a new web socket was created.

and in IE, the maximum websocket connections are 6 by default, https://msdn.microsoft.com/ja-jp/library/ee330736%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396.
so error occurs, you can change the number by modify the registry.

\HKEY_CURRENT_USER\Software\Policies\Microsoft\Internet Explorer\Main\FeatureControl\FEATURE_WEBSOCKET_MAXCONNECTIONSPERSERVER
and add a key, iexplore.exe（REG_DWORD) modify the value to 2~128.

I just modify to 50, and the error gone.

**Comment by kevinkuszyk** [comment]
Thanks @danail-vasilev, putting the fix in the `index.html` as suggested by @JiaLiPassion  [above](https://github.com/angular/zone.js/issues/933#issuecomment-351731730) seems to fix it.

**Comment by JiaLiPassion** [comment]
@jochenhebbrecht , the `flag` will try to catch all error when access the `event handler`. So it may also have impact in `Chrome/firefox` such as  `chrome/firefox extension code` involved. There was an issue in `firefox 52` which is try to access `event handler` in main page from firefox extension. With this flag, the issue was resolved. 

So if you have the specified issue, I may be able to help you to see whether the `flag` will help.

**Comment by jochenhebbrecht** [comment]
Hi @JiaLiPassion ,

Thanks for your very fast feedback! :-)
So I conclude it's OK to put the `flag` for any kind of browser. Any plans for taking this up in a next release? (so you don't need to add the line)

Jochen

**Comment by JiaLiPassion** [comment]
@jochenhebbrecht , because this `flag` will basically impact `IE/Edge` developer tools, so it will not be `true` by default ( set to true will have some performance impact for `addEventListener` call). you can set it to `true` if you have such kind of use case.

**Comment by jochenhebbrecht** [comment]
@JiaLiPassion what do you mean with: it will have impact on IE/Edge developer tools? It seems to be working fine when put to **true**? The developer tools still work?
In the second part of your reply, your stating putting the value to **true** will impact the performance. You mean: it will work more slowly when put to **true**? If that's the case, then the value should be **false** when rendering in Google Chrome/Mozilla Firefox.

**Comment by JiaLiPassion** [comment]
@jochenhebbrecht , yes, everything will work if you set it to `true`. enable the `flag` will cause an additional `try/catch` and `toString` operation when you call `addEventListener`. so it will be a little `performance impact` when you call `addEventListener`, the behavior is the same.

And the reason don't want to set it to `true` by default even in IE because it only impact some of the `developer tools` , so it only impact `development` mode, and in the new version of `angular/cli` generated `project`, you can find the following code in `src/polyfills.ts`.

```javascript

 /*
 * in IE/Edge developer tools, the addEventListener will also be wrapped by zone.js
 * with the following flag, it will bypass `zone.js` patch for IE/Edge
 */
// (window as any).__Zone_enable_cross_context_check = true;
``` 

you can open the comment when you in development mode and comment it when you want to build in production mode.

**Comment by jochenhebbrecht** [comment]
Thanks @JiaLiPassion !

**Comment by ojemuyiwa** [comment]
Thanks so much @JiaLiPassion 

Added only for IE and Edge using the below in pollyfills.ts and it works a charm
`/*
* in IE/Edge developer tools, the addEventListener will also be wrapped by zone.js
* with the following flag, it will bypass `zone.js` patch for IE/Edge
*/
//(window as any).__Zone_enable_cross_context_check = true;
if (document['documentMode'] || /Edge/.test(navigator.userAgent)) {
    (window as any).__Zone_enable_cross_context_check = true;
}`

**Comment by maxmumford** [comment]
I seem to get this error in IE11 while dev tools are open, regardless of whether or not JiaLiPassion's fix is applied. Closing dev tools makes the web app load again.

**Comment by mleibman** [comment]
Has this been documented?
This thread is a bit hard to parse, so please correct me if I'm wrong:

- This fixes an intermittent issue in IE11 that manifests **only** if the dev tools are open.
- Enabling it in production is not recommended since it has performance overhead.
- Enabling it during development or for unit tests is fine.

**Comment by stcoder** [comment]
Try to specify the version of IE in .browser list rc, helped me.

```
### Supported Browsers

> 1%
last 2 versions
iOS > 7
ie 6-11
```

**Comment by Deepakm1051** [comment]
@JiaLiPassion 

i have added the following code in index.html. It does not work for me still the application is keep loading. My developer tool was closed.

 <script>
    __Zone_enable_cross_context_check = true;
    </script>

i tried this approach also

In pollyfills.ts, uncommented the following line

(window as any).__Zone_enable_cross_context_check = true;

could you help me resolve this issue?

Thanks in advance.
